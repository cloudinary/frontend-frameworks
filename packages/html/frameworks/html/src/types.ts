import type {CloudinaryImage} from "@cloudinary/base/assets/CloudinaryImage";
import type {ITranscodeAction} from "@cloudinary/base/actions/transcode";

export type Plugin = (element: HTMLImageElement|HTMLVideoElement, cloudinaryImage: CloudinaryImage, htmlPluginState?: HtmlPluginState) => Promise<string | void>;

export type Plugins = Plugin[];

export type AccessibilityMode = 'darkmode'|'brightmode'|'monochrome'|'colorblind';

export type PlaceholderMode = 'vectorize' | 'pixelate' | 'blur' | 'predominant-color';

export type HtmlPluginState = { cleanupCallbacks: Function[], pluginEventSubscription: Function[] };

export type VideoSources = {type: VideoType, codecs: Array<string>, transcode: ITranscodeAction}[] | undefined;

export type VideoType = 'flv'|'3gp'|'mov'|'mpg'|'avi'|'wmv'|'ogv'|string;
