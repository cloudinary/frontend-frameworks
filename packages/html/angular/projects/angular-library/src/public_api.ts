/*
 * Public API Surface of angular-library
 */

export * from './lib/angular-library.service';
export * from './lib/angular-library.component';
export * from './lib/angular-library.module';
