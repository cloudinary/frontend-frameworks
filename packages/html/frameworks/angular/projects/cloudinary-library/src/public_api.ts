/*
 * Public API Surface of cloudinary-library
 */

import {
  placeholder,
  accessibility,
  lazyload,
  responsive,
} from '../../../../html/dist';
export {placeholder, accessibility, lazyload, responsive};

export * from './lib/cloudinary.module';
