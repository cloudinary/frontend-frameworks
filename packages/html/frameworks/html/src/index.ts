export { HtmlImageLayer } from './layers/htmlImageLayer';
export { HtmlVideoLayer } from './layers/htmlVideoLayer';
export { HtmlPictureLayer } from './layers/htmlPictureLayer';
export {responsive} from './plugins/responsive';
export {lazyload} from './plugins/lazyload';
export {accessibility} from './plugins/accessibility';
export {placeholder} from './plugins/placeholder';
export {isBrowser} from './utils/isBrowser';
export {serverSideSrc} from './utils/serverSideSrc';
export {Plugins, VideoSources, PictureSources} from './types';
export {cancelCurrentlyRunningPlugins} from './utils/cancelCurrentlyRunningPlugins';
