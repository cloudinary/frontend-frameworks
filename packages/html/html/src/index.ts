export { HtmlLayer } from './htmlLayer';
export {responsive} from './responsive';
export {lazyload} from './lazyload';
export {accessibility} from './accessibility';
export {placeholder} from './placeholder';
export {isBrowser} from './isBrowser';
export {serverSideSrc} from './serverSideSrc';
export {html} from './html'
export {plugins} from './htmlLayer';

