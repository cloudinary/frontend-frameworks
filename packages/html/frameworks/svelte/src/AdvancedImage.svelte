<script lang="ts">
  import { CloudinaryImage } from '@cloudinary/url-gen/assets/CloudinaryImage';
  import { afterUpdate, onDestroy } from 'svelte';
  import { HtmlImageLayer, isBrowser, serverSideSrc , cancelCurrentlyRunningPlugins} from '@cloudinary/html';
  import { Plugins } from '@cloudinary/html';

  /**
   * @mixin SvelteSDK
   * @description The Cloudinary Svelte SDK contains components like <AdvancedImage> to easily render your media assets from Cloudinary.
   * The SDK also comes with support for optional js plugins that make the components smart, with features like lazy loading, placeholder, accessibility & responsiveness
   *
   * @example
   * <caption>
   *  Please note that the order of the plugins is important. See home for more details.
   * </caption>
   * // Example
   *
   * <script>
   *
   * import {CloudinaryImage} from '@cloudinary/url-gen';
   * import { AdvancedImage, accessibility, responsive, lazyload, placeholder } from '@cloudinary/svelte';
   *
   * const img = new CloudinaryImage('sample', {cloudName: 'demo'});
   *
   * // Close script tag
   * </scr\ipt>
   *
   * <AdvancedImage cldImg={img} plugins={[lazyload(),responsive(), accessibility(), placeholder()]}/>
   *
   */

  // Props passed by user
  export let cldImg: CloudinaryImage; // Required prop
  export let plugins: Plugins = []; // Optional prop

  // Internal variables used by this component
  let imgElement: HTMLImageElement; // Reference to underlying <img> element
  let htmlLayerInstance: HtmlImageLayer; // Updates dom using given cldImg & plugins

  /**
   * Bind imgElement to the underlying <img> element.
   * @param node - the underlying <img> element's node.
   */
  const bindImage = (node) => (imgElement = node);

  /**
   * On mount: Create a new HTMLLayer instance
   * On props change: Cancel running plugins and update image instance
   */
  afterUpdate(() => {
    if (!htmlLayerInstance) {
      if (imgElement) {
        htmlLayerInstance = new HtmlImageLayer(imgElement, cldImg, plugins);
      }
    } else {
      cancelCurrentlyRunningPlugins(htmlLayerInstance.htmlPluginState);
      htmlLayerInstance.update(cldImg, plugins);
    }
  });

  /**
   * On unmount we cancel the currently running plugins
   * This is the only lifecycle method that runs on the server,
   * so we must check for existence of htmlLayerInstance
   */
  onDestroy(() => {
    if (htmlLayerInstance) {
      cancelCurrentlyRunningPlugins(htmlLayerInstance.htmlPluginState);
    }
  });
</script>

{#if isBrowser()}
    <!-- svelte-ignore a11y-no-img- -->
    <img use:bindImage {...$$restProps}/>
{:else}
    <img src={serverSideSrc(plugins, cldImg)} {...$$restProps}/>
{/if}
