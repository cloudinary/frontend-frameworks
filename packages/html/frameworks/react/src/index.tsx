import {
  placeholder,
  accessibility,
  lazyload,
  responsive,
} from '../../html/dist'
import {AdvancedImage} from "./AdvancedImage";


export {placeholder, accessibility, lazyload, responsive, AdvancedImage};
