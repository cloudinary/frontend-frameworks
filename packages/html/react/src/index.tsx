import * as React from 'react'
import {
  HtmlLayer,
  placeholder,
  accessibility,
  lazyload,
  responsive,
  cancelCurrentlyRunningPlugins,
  isBrowser,
  serverSideSrc
} from '../../html/dist'

export class CldImg extends React.Component {
  private myRef: React.RefObject<unknown>;
  private canceled: any[];
  private imageInstance: HtmlLayer;
  constructor(props: any) {
    super(props);
    this.myRef = React.createRef();
    this.state = {
      transformation: props.transformation,
      plugins: props.plugins
    }
  }

  /**
   * On mount creates a new HTMLLayer instance and initialises with ref to img element,
   * user generated transformableImg, the plugins to be used, and an empty array to hold state
   * of cancelled plugins
   */
  componentDidMount() {
    this.canceled = []; // holds 'canceled' resolves
    this.imageInstance = new HtmlLayer(
      this.myRef.current,
      // @ts-ignore
      this.state.transformation,
      // @ts-ignore
      this.state.plugins,
      this.canceled
    )
  }

  /**
   * On update we cancel running plugins and update image instance with the state of user
   * transformableImg, the state of plugins, and we reset the state of the canceled plugins
   */
  componentDidUpdate() {
    cancelCurrentlyRunningPlugins(this.canceled);
    // call render again with plugins and reset toBeCanceled
    // @ts-ignore
    this.imageInstance.update(this.state.transformation, this.state.plugins, [])
  }

  /**
   * On unmount we cancel the currently running plugins
   */
  componentWillUnmount() {
    // safely cancel running events on unmount
    cancelCurrentlyRunningPlugins(this.canceled)
  }

  render() {
    if (isBrowser()) {
      console.log('client');
      // @ts-ignore
      return <img suppressHydrationWarning {...this.props} ref={this.myRef} />
    } else {
      console.log('server');
      const src = serverSideSrc(
        '', // src
        // @ts-ignore
        this.state.plugins,
        // @ts-ignore
        this.props.transformation
      );
      return <img {...this.props} src={src} />
    }
  }
}

export {placeholder, accessibility, lazyload, responsive};

