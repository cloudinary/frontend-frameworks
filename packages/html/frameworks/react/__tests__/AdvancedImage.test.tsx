import { AdvancedImage } from '../src'
import {CloudinaryImage} from "@cloudinary/base/assets/CloudinaryImage";
import {mount} from 'enzyme';
import React from "react";

const cloudinaryImage = new CloudinaryImage('sample', { cloudName: 'demo'});

describe('AdvancedImage', () => {
  it('is truthy', () => {
    expect(AdvancedImage).toBeTruthy()
  });

  it('renders AdvancedImage', async () => {
    const component = await mount(<AdvancedImage cldImg={cloudinaryImage}/>);
    expect(component).toMatchSnapshot();
  });

  it("should create an img tag", async function() {
    let component = await mount(<AdvancedImage cldImg={cloudinaryImage}/>);
    expect(component.html()).toBe('<img src="https://res.cloudinary.com/demo/image/upload/sample">');
  });

  it("should add style to img component", async function() {
    let component = await mount(<AdvancedImage style={{opacity: "0.5"}} cldImg={cloudinaryImage}/>);
    expect(component.find('img').prop('style')).toStrictEqual({"opacity": "0.5"});
  });
});
