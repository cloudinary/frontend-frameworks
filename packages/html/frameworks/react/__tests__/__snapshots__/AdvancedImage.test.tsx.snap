// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdvancedImage renders AdvancedImage 1`] = `ReactWrapper {}`;
